From bc7dcac109f1727997e03ff8733317a8b12f08dd Mon Sep 17 00:00:00 2001
From: Satchin Campbell <satchin.campbell@gmail.com>
Date: Sun, 4 Dec 2016 10:42:48 -0500
Subject: [PATCH] Fixed bugs, issue 9 resolved

---
 free.asm | 9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/free.asm b/free.asm
index 16485e8..df0f010 100644
--- a/free.asm
+++ b/free.asm
@@ -4,12 +4,10 @@ main:
     jal free
     jr $ra
 free:
-    lw $a0, 0($sp)
-    addi $sp, $sp, 4
-    beq $a0, 0, nullptr
+    beq $a0, $0, nullptr
 #else
     #store address of start of node meta in t0
-    addi $t0, $a0, -12
+    addi $t0, $a0, -12 #a0 stores argument pointer to be freed
     addi $t1, $t0, 4 #previous address is stored here
     addi $t2, $t0, 8 #next address is stored here
     lw $t3, 0($t1)   #previous node
@@ -18,7 +16,6 @@ free:
     sw $t4, 0($t3) #update previous node next to current node next
     addi $t4, $t4, 4 #location of next node previous
     addi $t3, $t3, -8 #restore location of previous node next to previous node
-    sw $ $t3 , 0($t4) #store location of current node previous to next node previous
-
+    sw $t3 ,0($t4) #store location of current node previous to next node previous
 nullptr:
     jr $ra
-- 
2.1.4

